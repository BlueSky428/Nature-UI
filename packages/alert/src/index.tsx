import * as React from 'react';
import clx from 'clsx';
import { FiInfo, FiAlertCircle, FiAlertTriangle, FiCheckCircle } from 'react-icons/fi';

const SUBTLE_TEXT = 'text-gray-800';

export const ALERT_STATUSES = {
  error: {
    bg: 'bg-red-200',
    iconColor: 'text-red-600 mr-3',
    icon: FiAlertCircle,
    variant: {
      solid: 'bg-red-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  info: {
    bg: 'bg-blue-200',
    iconColor: 'text-blue-600 mr-3',
    icon: FiInfo,
    variant: {
      solid: 'bg-blue-600',
      subtle: SUBTLE_TEXT,
    },
  },
  success: {
    bg: 'bg-green-200',
    iconColor: 'text-green-600 mr-3',
    icon: FiCheckCircle,
    variant: {
      solid: 'bg-green-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  warning: {
    bg: 'bg-orange-200',
    iconColor: 'text-orange-600 mr-3',
    icon: FiAlertTriangle,
    variant: {
      solid: 'bg-orange-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
};

interface AlertOptions {
  /**
   * The status of the alert
   */
  status?: keyof typeof ALERT_STATUSES;
  /**
   * The variant of the alert style to use
   */
  variant?: 'solid' | 'subtle';

  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;

  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component?: React.ElementType;
}

export type AlertProps = AlertOptions;

const BASE_STYLE = 'px-4 py-3 flex items-center';

const Alert: React.ForwardRefExoticComponent<AlertProps> = React.forwardRef(
  (props: AlertProps, ref: React.Ref<any>) => {
    const {
      className = '',
      status = 'success',
      children,
      variant = 'subtle',
      component: Component = 'div',
    } = props;

    // const Component = 'div';

    const VARIANT: string = ALERT_STATUSES[status].variant[variant];

    const componentClass = clx(BASE_STYLE, {
      [className]: className,
      [ALERT_STATUSES[status].bg]: status,
      [VARIANT]: variant,
    });

    const Icon = ALERT_STATUSES[status];
    const iconClasses = clx({
      [VARIANT]: variant,
      'mr-3': variant,
      [Icon.iconColor]: variant !== 'solid',
    });

    const IconComponent = Icon.icon;

    return (
      <Component className={componentClass} ref={ref}>
        {status && <IconComponent className={iconClasses} size={20} />}
        <span>{children ? children : 'This is an alert ðŸ™‚'}</span>
      </Component>
    );
  }
);

Alert.displayName = 'Alert';

export default Alert;
