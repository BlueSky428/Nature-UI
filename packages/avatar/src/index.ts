export * from './avatar';
export * from './avatarGroup';
