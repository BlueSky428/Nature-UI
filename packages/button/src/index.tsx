import * as React from 'react';
import clx from 'clsx';

interface ButtonProps {
  /**
   * The label of the button
   */
  'aria-label'?: string;
  /**
   * The text color of the button. Use a color key passed in theme.colors.
   */
  text?: string;
  /**
   * The background color of the button. Use a color key passed in theme.colors.
   */
  color?: string;
  /**
   * The variant of the button style to use.
   */
  variant?: 'outline' | 'ghost' | 'link' | 'solid' | 'none';
  /**
   * If true, the button will be disabled.
   */
  isDisabled?: boolean;
  /**
   * If true, the button will show a spinner.
   */
  isLoading?: boolean;
  /**
   * The label to show in the button when isLoading is true. If no text is passed, it only shows the spinner
   */
  loadingText?: string;
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'none';
  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;
  /**
   * The lightness or how dark you want the background color to be
   */
  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   * defaults to "button"
   */
  component?: React.ElementType;
  id?: string;
  /**
   *  when the href prop is set it changes the tag name to a
   */
  href?: string;
  onClick?: () => any;
}

const darken = (value: string): string => {
  const splitStr = value.split('-');
  const dept = Number(splitStr[1]);
  const color = splitStr[0];

  if (dept >= 800) {
    return `${color}-${dept - 200}`;
  }

  return `${color}-${dept + 200}`;
};

const lighten = (value: string): string => {
  const color = value.split('-')[0];

  return `${color}-${100}`;
};

const Button = React.forwardRef((props: ButtonProps, ref: React.Ref<any>) => {
  const {
    component: Component = 'button',
    variant = 'solid',
    color = 'teal-500',
    text = 'white',
    size = 'md',
    children,
    ...rest
  } = props;
  const DEFAULT_CLASS = 'block focus:shadow-outline focus:outline-none rounded font-semibold flex';
  const STYLES = {
    solid: `bg-${color} text-${text} hover:bg-${darken(color)}`,
    outline: `bg-transparent hover:bg-${lighten(
      text
    )} text-${text} border border-${text} focus:border-transparent`,
    ghost: `hover:bg-${lighten(text)} text-${text}`,
    link: `hover:underline text-${text}`,
    md: 'px-4 py-2',
  };

  let BTNClass: string;

  if (variant === 'none') {
    BTNClass = '';
  } else {
    BTNClass = clx(DEFAULT_CLASS, {
      [STYLES[size]]: size && variant !== 'link',
      [STYLES[variant]]: variant,
    });
  }

  return (
    <Component className={BTNClass} ref={ref} {...rest}>
      {children ? children : 'Button'}
    </Component>
  );
});

export default Button;
