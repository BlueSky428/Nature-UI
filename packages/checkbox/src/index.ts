export * from './checkbox';
export * from './checkbox-group';
export * from './checkbox-types';
export * from './use-checkbox';
export * from './use-checkbox-group';
