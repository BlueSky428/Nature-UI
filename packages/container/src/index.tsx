import * as React from 'react';
import clx from 'clsx';

interface IContainer {
  children?: React.ReactNode;
  /**
   * Determine the max-width of the container.
   */
  size?: 'lg' | 'md' | 'sm' | 'xl' | 'xs';
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component?: React.ComponentType;
  className?: string;
}

const BASE_STYLE = 'mx-auto container';
const LG = 'px-20';
const MD = 'px-32';
const SM = 'px-48';
const XL = 'px-0';
const XS = 'px-64';

const Container: React.ForwardRefExoticComponent<IContainer> = React.forwardRef(
  (props: IContainer, ref: React.Ref<any>) => {
    const { component: Component = 'div', size = 'lg', children, className = '' } = props;

    const componentClass: string = clx(BASE_STYLE, {
      [LG]: size === 'lg',
      [MD]: size === 'md',
      [SM]: size === 'sm',
      [XL]: size === 'xl',
      [XS]: size === 'xs'
    });

    return (
      <Component className={componentClass} ref={ref}>
        <Component className={className}>{children}</Component>
      </Component>
    );
  }
);

Container.displayName = 'Container';

export default Container;
