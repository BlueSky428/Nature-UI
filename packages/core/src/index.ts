export * from '@nature-ui/alert';
export * from '@nature-ui/alert-dialog';
export * from '@nature-ui/avatar';
export * from '@nature-ui/breadcrumb';
export * from '@nature-ui/button';
export * from '@nature-ui/checkbox';
export * from '@nature-ui/close-button';
export * from '@nature-ui/container';
export * from '@nature-ui/counter';
export * from '@nature-ui/drawer';
export * from '@nature-ui/form-control';
export * from '@nature-ui/hooks';
export * from '@nature-ui/icon';
export * from '@nature-ui/image';
export * from '@nature-ui/input';
export * from '@nature-ui/layout';
export * from '@nature-ui/modal';
export * from '@nature-ui/popover';
export * from '@nature-ui/popper';
export * from '@nature-ui/portal';
export * from '@nature-ui/progress';
export * from '@nature-ui/radio';
export * from '@nature-ui/spinner';
export * from '@nature-ui/system';
export * from '@nature-ui/toast';
export * from '@nature-ui/transition';
export * from '@nature-ui/visually-hidden';
