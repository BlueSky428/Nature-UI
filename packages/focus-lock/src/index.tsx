import * as React from 'react';
import { getAllFocusable, isFocusable, __DEV__ } from '@nature-ui/utils';
import ReactFocusLock from 'react-focus-lock';

export interface FocusLockProps {
  /**
   * `ref` of the element to receive focus initially
   */
  initialFocusRef?: React.RefObject<HTMLElement>;
  /**
   * `ref` of the element to return focus to when `FocusLock`
   * unmounts
   */
  finalFocusRef?: React.RefObject<HTMLElement>;
  /**
   * The `ref` of the wrapper for which the focus-lock wraps
   */
  contentRef?: React.RefObject<HTMLElement>;
  /**
   * If `true`, focus will be restored to the element that
   * triggered the `FocusLock` once it unmounts
   */
  restoreFocus?: boolean;
  /**
   * The component to render
   */
  children: React.ReactNode;
  /**
   * If `true`, focus trapping will be disabled
   */
  isDisabled?: boolean;
  /**
   * If `true`, the first focuable element within the `children`
   * will ne auto-focused once `FocusLock` mounts
   */
  autoFocus?: boolean;
}

/**
 * React component to trap focus within an element or component.
 * Mostly used in Modals, Popovers, etc.
 */
export const FocusLock = (props: FocusLockProps) => {
  const {
    initialFocusRef,
    finalFocusRef,
    contentRef,
    restoreFocus,
    children,
    isDisabled,
    autoFocus = true,
  } = props;

  const onActivation = React.useCallback(() => {
    if (initialFocusRef?.current) {
      initialFocusRef.current.focus();
    } else {
      if (contentRef?.current) {
        const focusable = getAllFocusable(contentRef.current);
        if (focusable.length === 0) {
          contentRef?.current?.focus();
        }
      }
    }
  }, [initialFocusRef, contentRef]);

  const onDeactivation = React.useCallback(() => {
    finalFocusRef?.current?.focus();
  }, [finalFocusRef]);

  const returnFocus = restoreFocus && !finalFocusRef;

  return (
    <ReactFocusLock
      autoFocus={autoFocus}
      disabled={isDisabled}
      onActivation={onActivation}
      onDeactivation={onDeactivation}
      returnFocus={returnFocus}
    >
      {children}
    </ReactFocusLock>
  );
};

if (__DEV__) {
  FocusLock.displayName = 'FocusLock';
}

export default FocusLock;
