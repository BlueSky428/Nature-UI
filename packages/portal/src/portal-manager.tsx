import { createContext } from '@nature-ui/react-utils';
import { __DEV__ } from '@nature-ui/utils';
import React from 'react';

interface PortalManagerContext {
  zIndex?: number;
}

const [PortalManagerContextProvider, usePortalManager] =
  createContext<PortalManagerContext | null>({
    strict: false,
    name: 'PortalManagerContext',
  });

export { usePortalManager };

export interface PortalManagerProps {
  /**
   * Child elements of the Portal manager
   * Ideally, it should be at the top-level
   * of your application
   */
  children?: React.ReactNode;
  /**
   * [Z-Index war] If your has multiple elements
   * with z-index clashing, you might need to
   * apply a z-index to the Portal manager
   */
  zIndex?: number;
}

/**
 * PortalManager
 *
 * Used to manage multiple portals within an application.
 * It must be render only once, at the root of your application.
 *
 * Inspired by BaseWeb's LayerManager component
 */
export const PortalManager = (props: PortalManagerProps) => {
  const { children, zIndex } = props;
  return (
    <PortalManagerContextProvider value={{ zIndex }}>
      {children}
    </PortalManagerContextProvider>
  );
};

if (__DEV__) {
  PortalManager.displayName = 'PortalManager';
}
