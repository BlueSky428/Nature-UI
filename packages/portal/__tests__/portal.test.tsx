import * as React from 'react';
import { render } from '@nature-ui/test-utils';

import { PortalManager, Portal } from '../src';

describe('@nature-ui/portal', () => {
  test('should render portal', () => {
    const tools = render(
      <PortalManager>
        <Portal>This is a portal</Portal>
      </PortalManager>,
    );

    expect(tools.baseElement).toMatchSnapshot();
  });

  test('should render nested portal', () => {
    const tools = render(
      <PortalManager>
        <Portal>
          This is a portal.
          <Portal>This is a nested portal</Portal>
        </Portal>
      </PortalManager>,
    );

    expect(tools.baseElement).toMatchSnapshot();

    const portals = Array.from(
      tools.baseElement.querySelectorAll('.nature-portal'),
    );

    const [parentPortal, childPortal] = portals;
    expect(parentPortal).toContainElement(childPortal as HTMLElement);
  });

  test('should render into a custom container', () => {
    const Custom = () => {
      const ref = React.useRef<any>(null);
      return (
        <PortalManager>
          <div data-testid='container' ref={ref} />
          <Portal getContainer={() => ref.current}>
            <h1 data-testid='heading'>Hello world</h1>
          </Portal>
        </PortalManager>
      );
    };

    const tools = render(<Custom />);

    expect(tools.baseElement).toMatchSnapshot();

    const heading = tools.getByTestId('heading');
    const container = tools.getByTestId('container');

    expect(container).toContainElement(heading);
  });
});
