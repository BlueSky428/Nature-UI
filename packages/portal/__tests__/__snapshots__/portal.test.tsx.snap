// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nature-ui/portal should render into a custom container 1`] = `
<body>
  <div>
    <div
      data-testid="container"
    >
      <div
        class="nature-portal"
      >
        <h1
          data-testid="heading"
        >
          Hello world
        </h1>
      </div>
    </div>
    <div />
  </div>
</body>
`;

exports[`@nature-ui/portal should render nested portal 1`] = `
<body>
  <div>
    <div>
      <div
        class="nature-portal"
      >
        This is a portal.
        <div
          class="nature-portal"
        >
          This is a nested portal
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`@nature-ui/portal should render portal 1`] = `
<body>
  <div>
    <div>
      <div
        class="nature-portal"
      >
        This is a portal
      </div>
    </div>
  </div>
</body>
`;
